Implement Tombstone Message Handling for JDBC Sink Connector

Joel-hanson opened the pull request:
## Description:

This pull request addresses issue #165, focusing on enhancing the functionality of the JDBC sink connector to handle deletes on tombstone messages effectively. By implementing this feature, users can now delete rows corresponding to tombstone messages, which is particularly useful for scenarios involving Change Data Capture (CDC) from another database.

#### Changes:

- Added support for handling tombstone messages in the JDBC sink connector.
- Implemented the ability to delete rows based on tombstone messages.
- Introduced a new parameter, `delete.enabled`, to control delete behavior.
- Aligned functionality with the documented approach for processing tombstones, similar to Confluent JDBC driver behavior.

**Related Issue(s):**
- Fixes #165


Signed-off-by: Joel Hanson <joelhanson025@gmail.com>

Joel-hanson commented:
@ivanyu, could you please take a look at this pull request when you get a chance? Your feedback would be really valuable. Thank you!

matuagarwal commented:
@jeqo @ivanyu could you please review the PR

davidradl commented on src/main/java/io/aiven/connect/jdbc/sink/BufferedRecords.java:
```diff
@@ -109,25 +118,32 @@ public List<SinkRecord> add(final SinkRecord record) throws SQLException {
 
     private void prepareStatement() throws SQLException {
-        final String sql;
-        log.debug("Generating query for insert mode {} and {} records", config.insertMode, records.size());
```
Why are we deleting this debug line. It seems like it would be useful.

davidradl commented on src/main/java/io/aiven/connect/jdbc/sink/BufferedRecords.java:
```diff
@@ -153,14 +169,14 @@ private void reInitialize(final SchemaPair schemaPair) throws SQLException {
-        if (records.isEmpty()) {
-            log.debug("Records is empty");
+        if (records.isEmpty() && tombstoneRecords.isEmpty()) {
+            log.debug("Records are empty.");
```
nit
I am not sure about this debug.
- the debug text should include Records and tombstone records are empty. 
- Do we  want to have a debug is the records are empty as before and there are tombstones .

Joel-hanson replied:
> Do we want to have a debug is the records are empty as before and there are tombstones.

It seems redundant to include if statements solely for debugging empty records, as the prepare statement debug (mentioned previously) will already cover this aspect.

davidradl commented on src/main/java/io/aiven/connect/jdbc/sink/BufferedRecords.java:
```diff
@@ -86,10 +90,15 @@ public List<SinkRecord> add(final SinkRecord record) throws SQLException {
+                records.add(record);
+            }
+            if (records.size() >= config.batchSize || tombstoneRecords.size() >= config.batchSize) {
                 log.debug("Flushing buffered records after exceeding configured batch size {}.",
```
nit
might be worth adding in the records.size() and tombstoneRecords.size() so we know which one caused the flush

Joel-hanson replied:
Thanks @davidradl for the review, I have addressed most of the comments

aindriu-aiven commented:
Hey @Joel-hanson, thanks a million for your contribution I have this on my list for review on Monday, sorry for the delay I am just catching up on a few projects at the moment!

C0urante commented on src/main/java/io/aiven/connect/jdbc/sink/BufferedRecords.java:
```diff
@@ -169,25 +185,40 @@ public List<SinkRecord> flush() throws SQLException {
         }
         log.debug("Done executing batch.");
-        if (totalUpdateCount != records.size() && !successNoInfo) {
+        verifySuccessfulExecutions(totalSuccessfulExecutionCount, successNoInfo);
```
Should we invoke this once per executed batch (i.e., once for regular records and once for tombstone records)? Otherwise, the `successNoInfo` field gets reused across batches and may produce inaccurate results.

C0urante commented on src/main/java/io/aiven/connect/jdbc/sink/JdbcSinkConfig.java:
```diff
@@ -387,6 +407,17 @@ public JdbcSinkConfig(final Map<?, ?> props) {
+            // requires the primary key be used as criteria.
+            deleteEnabled = getBoolean(DELETE_ENABLED);
+        } else {
+            if (getBoolean(DELETE_ENABLED)) {
+                log.error("Delete mode will enabled only if pk mode set to record_key");
+            }
```
IMO we should fail here instead of just logging an error. Bonus points if we implement this multi-property validation in the [Connector::validate method](https://kafka.apache.org/37/javadoc/org/apache/kafka/connect/connector/Connector.html#validate(java.util.Map)), but I don't think it's necessary to go that far for this PR if you don't have the time.

C0urante reviewed (commented):
Thanks @Joel-hanson for the PR! I've left a few thoughts but overall it looks pretty good.

I'm unable to run the integration tests on my Apple Silicon Macbook, though. I've tried several things (including looking through Testcontainers issues and using an alternative Docker runtime) with no success.

Would it be possible to use Postgres either instead of or in addition to Oracle for the new integration tests? As Apple Silicon becomes more and more prevalent for devs the cost of having tests that can't be run locally increases.

Joel-hanson commented:
@aindriu-aiven @C0urante I have addressed most of the comments. Could you please do a re-review?

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/oracle/VerifyInsertIT.java:
```diff
@@ -0,0 +1,189 @@
+    public void testSinkConnector() throws Exception {
+        createTopic(TEST_TOPIC_NAME, 1); // Create Kafka topic matching the table name
+        executeSqlStatement(CREATE_TABLE);
+        consumer.assign(Collections.singleton(new TopicPartition(TEST_TOPIC_NAME, 0)));
```
Don't think this is necessary here or in the other tests in this suite; we never use the consumer.

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/oracle/VerifyInsertIT.java:
```diff
@@ -0,0 +1,189 @@
+        executeSqlStatement(CREATE_TABLE);
+        consumer.assign(Collections.singleton(new TopicPartition(TEST_TOPIC_NAME, 0)));
+        // Start the sink connector
+        final Map<String, String> config = basicSinkConnectorConfig();
+        config.put("pk.mode", "record_key");
+        config.put("pk.fields", "id"); // assigned name for the primitive key
```
We can simplify this, right?
```suggestion
        final Map<String, String> config = sinkConnectorConfigWithPKModeRecordKey();
```

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/postgres/VerifyInsertIT.java:
```diff
@@ -0,0 +1,208 @@
+    public void testSinkConnector() throws Exception {
+        createTopic(TEST_TOPIC_NAME, 1); // Create Kafka topic matching the table name
+        executeSqlStatement(CREATE_TABLE);
+        consumer.assign(Collections.singleton(new TopicPartition(TEST_TOPIC_NAME, 0)));
```
Don't need these lines in this test suite either.

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/oracle/VerifyInsertIT.java:
```diff
@@ -0,0 +1,189 @@
+        // Send test data to Kafka topic
+        sendTestData(1);
+
+        await().atMost(Duration.ofSeconds(20)).pollInterval(Duration.ofSeconds(19))
```
Just curious, why specify a poll interval of 19 seconds? Wouldn't we want to poll more frequently like once a second in order for the assertion (and, as a result, the tests) to pass sooner?

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/postgres/VerifyInsertIT.java:
```diff
@@ -0,0 +1,208 @@
+        executeSqlStatement(CREATE_TABLE);
+        consumer.assign(Collections.singleton(new TopicPartition(TEST_TOPIC_NAME, 0)));
+        // Start the sink connector
+        final Map<String, String> config = basicSinkConnectorConfig();
+        config.put("pk.mode", "record_key");
+        config.put("pk.fields", "id"); // assigned name for the primitive key
```
Can also simplify this part:
```suggestion
        final Map<String, String> config = sinkConnectorConfigWithPKModeRecordKey();
```

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/oracle/VerifyDeleteIT.java:
```diff
@@ -0,0 +1,317 @@
+        return config;
+    }
+
+    private void assertNotEmptyPoll(final Duration duration) {
+        final ConsumerRecords<?, ?> records = consumer.poll(duration);
+        assertNotSame(ConsumerRecords.empty(), records);
+    }
```
What's the purpose of this method? Doesn't it just verify that some data has landed in Kafka, which we should already be reasonably sure of with our tracking of producer futures in `sendTestData` and similar methods?

C0urante replied:
(And if this method is unnecessary, we can probably also remove all calls to `consumer::assign` from these tests.)

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/oracle/VerifyDeleteIT.java:
```diff
@@ -0,0 +1,317 @@
+
+        sendTestDataWithTombstone(2);
+
+        await().atMost(Duration.ofSeconds(20)).pollInterval(Duration.ofSeconds(18))
+                .untilAsserted(() -> {
+                    assertThat(new Table(getDatasource(), TEST_TOPIC_NAME)).hasNumberOfRows(0);
+                });
```
This doesn't necessarily prove what we need it to. It's possible that we send the data, then make this assertion before the connector has had a chance to read it from Kafka and/or query the database.

If we were testing against a later version of Kafka Connect, we could wait for the connector to commit offsets (which can be verified via the `GET /connectors/{name}/offsets` endpoint introduced in version 3.5.0 as part of [KIP-875](https://cwiki.apache.org/confluence/display/KAFKA/KIP-875%3A+First-class+offsets+support+in+Kafka+Connect)).

But it's not worth adding all of that to this PR. Perhaps we could use `await().atLeast` instead of `await().atMost`, as a small temporary workaround?

C0urante replied:
This is still unaddressed. Let me know if you have any questions!

Joel-hanson replied:
I tried this, but when converting it to atLeast I faced an issue where the condition is fulfilled before the duration specified in the atLeast method. I don’t think it’s possible to know the exact duration for the test to fulfill, as it depends on the time it takes for the container to come up (which depends on the available resources) and start sending messages. Please correct me if my understanding is wrong.

C0urante replied:
Ah whoops, got the API wrong. Didn't realize it'd throw an exception if the condition were met before the `atLeast` duration expired.

As far as knowing the exact duration to wait goes--yes, you're correct that it's difficult to know exactly how long to wait. And with tombstone-only cases, if the connector is implemented correctly, then the table will have zero rows throughout the entire test.

This is why I brought up checking for committed offsets as a way to verify that the connector had consumed records from Kafka and sent them to the database (via either an insert or a delete query), but again, since it's probably not worth the work it'd take to implement that logic here, we can just wait a fixed interval and then verify that the table still has zero rows.

I'm sure there's a way to do this with Awaitility, though honestly at this point I don't know if that's really necessary. We could abandon Awaitility altogether and do something like this for the tombstone-only cases:
```java
// TODO: Instead of sleeping for a fixed interval, wait for the connector to commit offsets for the records we sent
Thread.sleep(5_000); // Give the connector at least five seconds to read our tombstone messages from Kafka
assertThat(new Table(getDatasource(), TEST_TOPIC_NAME)).hasNumberOfRows(0);
```

Joel-hanson replied:
Thanks for the clarification and suggestions

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/oracle/VerifyDeleteIT.java:
```diff
@@ -0,0 +1,317 @@
+    }
+
+    @Test
+    public void testMultiInsertMode() throws Exception {
```
We don't actually send any tombstone records in this test. Maybe it would make more sense in the `VerifyInsertIT` suite, similar to the `testSinkConnectorInDeleteMode` case that's already in that suite?

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/oracle/VerifyDeleteIT.java:
```diff
@@ -0,0 +1,317 @@
+
+        sendTestDataWithTombstone(2);
+
+        await().atMost(Duration.ofSeconds(20)).pollInterval(Duration.ofSeconds(18))
```
Same thought RE `atLeast` instead of `atMost`

C0urante replied:
Still unaddressed

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/postgres/VerifyDeleteIT.java:
```diff
@@ -0,0 +1,317 @@
+import static org.awaitility.Awaitility.await;
+import static org.junit.Assert.assertNotSame;
+
+public class VerifyDeleteIT extends AbstractPostgresIT {
```
I believe all the comments I left on the Oracle `VerifyDeleteIT` suite apply here as well; LMK if anything doesn't transfer, though.

C0urante reviewed (commented):
Thanks @Joel-hanson! This is looking really good. I've given all the changes to the main codebase and integration tests a pass. Still haven't looked at the unit tests yet; will try to do that in the next pass once these comments are addressed.

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/oracle/VerifyDeleteIT.java:
```diff
@@ -0,0 +1,276 @@
+        }
+    }
+
+    @BeforeEach
+    public void afterEach() throws SQLException {
```
We should remove this, and revert the change to make the `@Contaner`-annotated fields `static`; otherwise, the container is reused across test cases ([ref](https://java.testcontainers.org/test_framework_integration/junit_5/)).

I understand that it may be more efficient and faster to share containers across test methods, but it's not a requirement for this PR and right now some of our integration tests are failing if more than one is run in the same build.

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/oracle/AbstractOracleIT.java:
```diff
@@ -0,0 +1,77 @@
+            DockerImageName.parse("gvenzl/oracle-free")
+                    .withTag(DEFAULT_ORACLE_TAG);
+    @Container
+    public static final OracleContainer ORACLE_CONTAINER = new OracleContainer(DEFAULT_ORACLE_IMAGE_NAME);
```
This should match the convention set in `AbstractPostgresIT`:

1. Should not be `static` (otherwise, gets reused across tests)
2. Visibility should be `protected` instead of `public`
3. Name shouldn't be all-caps (due to being non-static)

```suggestion
    protected final OracleContainer oracleContainer = new OracleContainer(DEFAULT_ORACLE_IMAGE_NAME);
```

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/oracle/VerifyInsertIT.java:
```diff
@@ -0,0 +1,216 @@
+    }
+
+    @Test
+    public void testMultiInsertMode() throws Exception {
```
This name is a little misleading:
```suggestion
    public void testMultiInsertInDeleteMode() throws Exception {
```

C0urante commented on src/integrationTest/java/io/aiven/kafka/connect/jdbc/postgres/VerifyInsertIT.java:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    @Test
+    public void testMultiInsertMode() throws Exception {
```
```suggestion
    public void testMultiInsertInDeleteMode() throws Exception {
```

C0urante reviewed (dismissed):
LGTM, thanks @Joel-hanson!

One last request--can you squash your commits into a single commit? No need to mention "address review comments" or anything like that, we just need a description that'll be helpful in the git history for the master branch after the merge.

Joel-hanson commented:
> LGTM, thanks @Joel-hanson!
> 
> One last request--can you squash your commits into a single commit? No need to mention "address review comments" or anything like that, we just need a description that'll be helpful in the git history for the master branch after the merge.

@C0urante Thanks, I have squashed the commit and updated the documentation file to include the `delete.enabled` configuration.

Joel-hanson commented:
@C0urante Thank you for merging PR. I was wondering if you could provide an estimate of when the next release will be published that includes these changes?

If there's anything I can do to assist with the release process, please let me know.

Thank you again for your support and for maintaining this project.

C0urante commented:
@Joel-hanson I've reached out to some people internally, will let you know when I hear back.
